### DISCUSSION TITLE
Bot joins the channel but plays no sound. Help?

### DISCUSSION THREAD

**Discussion by kelly1266** [Rapptz/discord.py#9652] [Q&A] [answered]
Hi, I'm working on a bot that will join a voice channel and play a short sound byte whenever a non bot user enters a channel. The bot does join the channel and stays there for the length of the sound byte, however it does not produce any sound. Here is my code:

```
import discord
from discord.ext import commands
import time
from dotenv import load_dotenv
import os

load_dotenv()

intents = discord.Intents.all()
client = commands.Bot(command_prefix='!', intents=intents)

@client.event
async def on_ready():
    print(f'Logged in as {client.user}')

@client.event
async def on_voice_state_update(member, before, after):
    joined = (before.channel is None) and (after.channel is not None)
    valid_join_condition = (not member.bot) and (len(client.voice_clients) == 0)

    if (not valid_join_condition):
        return

    if (joined):
        ffmpeg_executable_path = os.environ.get("FFMPEG_PATH")
        mp3_file_path = "./intros/default.mp3"

        # Join channel and play the intro
        vc = await after.channel.connect()
        src = discord.FFmpegPCMAudio(executable=ffmpeg_executable_path, source=mp3_file_path)
        vc.play(src)
        while vc.is_playing():
            time.sleep(1)
        await vc.disconnect()

token = os.environ.get("DISCORD_TOKEN")
client.run(token)
```

Here is what I've tried so far:
- My headset is turned on and Discord has the correct output device set
- I have downloaded the version of discord.py with full voice support and have tried reinstalling PyNaCl
- I have enabled all privileged gateway intents (presence, server members, and message content)
- The bot has administrator access (as well as Speak, Use Voice Activity, and Connect permissions)
- I have downloaded the latest build of ffmpeg following the instructions laid out on this page (have tried both essentials and full builds)
    -  https://github.com/adaptlearning/adapt_authoring/wiki/Installing-FFmpeg
- My FFMPEG_PATH environment variable is pointing to the correct /bin/ffmpeg.exe file
- The default.mp3 file is in the correct location and has audio, i.e. its not 4 seconds of silence

Any suggestions as to what else I could check? Im sure I missed something, but not sure what it could be

**Comment by tailoric** [comment, marked as answer]
1. You should not `time.sleep(1)` ever in your bot. Discord.py is an async library and therefore if you need to wait something use `await asyncio.sleep()` instead [See here](https://discordpy.readthedocs.io/en/latest/faq.html?highlight=blocking#what-does-blocking-mean)
2. You don't need to do a waiting loop, play has an [`after`](https://discordpy.readthedocs.io/en/latest/api.html?highlight=after#discord.VoiceClient.play) keyworded argument that will call a function upon completion of the file.  You could do the disconnect in the `after` function. [See here](https://discordpy.readthedocs.io/en/latest/faq.html?highlight=after#how-do-i-pass-a-coroutine-to-the-player-s-after-function) for how to do it with an async function 

I am not 100% sure if this is the solution to your problems but it could be. At least blocking the bot for the duration of the audio file should not happen. Otherwise try to update to the current git version since there have been a few fixes related to voice functionality added to master which aren't in the latest release yet

`python -m pip install -U git+https://github.com/Rapptz/discord.py` should install the dev version.

**Reply by kelly1266** [reply]
Thank you, this solved my issue! The time library seems to have been the root of the problem, just removing the import and replacing time.sleep() solved the issue. Appreciate the help!
